commit 6717d9afef42
Author: Jean-Yves Avenard <jyavenard@mozilla.com>
Date:   Wed Jun 20 19:24:34 2018 +0000

    Bug 1469257 - [H264] Only check for SPS changes on keyframe. r=bryce, a=RyanVM
    
    Some invalid streams contain SPS changes and those appear to only occur on non-keyframe, this cause all frames to be dropped until the next keyframe is found. This result in apparent freezes.
    
    While it is theoretically possible to have SPS changes inband on non-keyframe those should be very rare (I've never seen one). The content would have been invalid anyway in an non-fragmented mp4.
    
    So we now only check for a SPS change on keyframe. This would cause no affect on either windows, android or ffmpeg as those decoders handle format change fine. The mac decoder could however show garbled frames temporarily.
    
    Differential Revision: https://phabricator.services.mozilla.com/D1733
    
    --HG--
    extra : source : ddde6ad6d9e6c4069632e195613bcd53a39883e4
---
 dom/media/platforms/wrappers/H264Converter.cpp | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git dom/media/platforms/wrappers/H264Converter.cpp dom/media/platforms/wrappers/H264Converter.cpp
index 32b1abf179937..5c04c075664c3 100644
--- dom/media/platforms/wrappers/H264Converter.cpp
+++ dom/media/platforms/wrappers/H264Converter.cpp
@@ -392,7 +392,7 @@ MediaResult
 H264Converter::CheckForSPSChange(MediaRawData* aSample)
 {
   RefPtr<MediaByteBuffer> extra_data =
-    mp4_demuxer::H264::ExtractExtraData(aSample);
+    aSample->mKeyframe ? mp4_demuxer::H264::ExtractExtraData(aSample) : nullptr;
   if (!mp4_demuxer::H264::HasSPS(extra_data)) {
     MOZ_ASSERT(mCanRecycleDecoder.isSome());
     if (!*mCanRecycleDecoder) {
@@ -406,14 +406,12 @@ H264Converter::CheckForSPSChange(MediaRawData* aSample)
     // This scenario can only occur on Android with devices that can recycle a
     // decoder.
     if (!mp4_demuxer::H264::HasSPS(aSample->mExtraData) ||
-        mp4_demuxer::H264::CompareExtraData(aSample->mExtraData,
-                                            mOriginalExtraData)) {
+        mp4_demuxer::H264::CompareExtraData(aSample->mExtraData, mOriginalExtraData)) {
       return NS_OK;
     }
     extra_data = mOriginalExtraData = aSample->mExtraData;
   }
-  if (mp4_demuxer::H264::CompareExtraData(extra_data,
-                                          mCurrentConfig.mExtraData)) {
+  if (mp4_demuxer::H264::CompareExtraData(extra_data, mCurrentConfig.mExtraData)) {
     return NS_OK;
   }
 
